### DISCUSSION TITLE
Whats the best way to process only selected pages from PDF

### DISCUSSION THREAD

**Discussion by AnweshGangula** [pymupdf/PyMuPDF#1283] [Q&A] [answered]
I'm working on an app that needs to add highlights, links, extract text and process. But most of the times this is done only for a few selected pages from a large pdf. So what i'm currently doing is this:

```python
doc = fitz.open(pdfPath)
selectedPages= [3,4,5,10,13,14,18]
for page in doc:
    if page.number in selectedPages:
    # do something

doc.save(newFile garbage=4, deflate=True, clean=True)
```

But, even though only few pages are selected, this method still need to write all other pages to the `newFile`, which makes sense. But it takes time to write all other pages to a new file. often longer than it takes to process the selectedPages because the PDF are of high quality.
What I'm looking for is - Is this the best way to approach this?

## Is there a way to only process the selectedPages and quickly replace those pages in the original PDF and save it as newFile?

For Example, one the features in "Blubeam Revu" - a PDF editor is to replace selected pages with another PDF, which is like this

1. Right-Click > Replace pages
2. Select a PDF to replace the existing pages with
3. Input a page-range to replace the pages in the original document
4. Enter to replace old pages with new one.
5. Save it as a new file or overwrite existing file

So in this above process, we are only editing the selected pages, which reduces the overall time it takes to save the PDF.

**Comment by JorjMcKie** [comment, marked as answer]
If you save to a new file, then obviously all page must be written. This is the same for all tools.

There are a few things you can do:
1. **_Do not_** write to a new file, instead use incremental save to update the existing file. Obviously this means the old pages' contents are still physcally there, however no longer accessible. This is the fastest alternative.
2. Play with the garbage collection level: garbage=4 is very, **_very_** time consuming, because it checks if there is any duplicate binary object content. That probably makes no sense at all in your case. Try `garbage=3`  or even `garbage=1`.
3. You probably also need no page cleaning for every page: omit `clean=True`.

**Reply by AnweshGangula** [reply]
Thank you @JorjMcKie, this is what I was looking for. 

And for people looking for this:

- `Document.saveIncr()` -  is a method _(convenience abbreviation)_ specifically to [save incrementally](https://pymupdf.readthedocs.io/en/latest/document.html#Document.saveIncr)
   - and here is the link to [wiki-page for incremental refresh](https://github.com/pymupdf/PyMuPDF/wiki/Using-Incremental-Saves)
- [detailed explanation](https://github.com/pymupdf/PyMuPDF/issues/655#issuecomment-691580281) on what `garbage` does

**Reply by AnweshGangula** [reply]
one more thing @JorjMcKie, the [documentation for doc.save()](https://pymupdf.readthedocs.io/en/latest/document.html#Document.save) states that positive value of garbage excludes incremental

> Positive values exclude “incremental”

Does this mean, incremental is not supported when garbage is positive?

**Reply by JorjMcKie** [reply]
yes!

**Reply by JorjMcKie** [reply]
Read the details in the docu: Incremental means append anything that has been changed to the file and redirect object pointers to the new stuff.
The complete previous file remains as it was before. So obviously there can be no garbage collection. Or any other structural changes (linear, etc.).
